website/notifications/utils.py

```diff
@@ -0,0 +1,160 @@
+import collections
+from framework.auth.signals import contributor_removed, node_deleted
+from website import settings
+from website.models import Node
+from website.notifications.model import Subscription
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+
+
+class NotificationsDict(dict):
+    def __init__(self):
+        super(dict, self).__init__()
+        self.update(messages=[], children=collections.defaultdict(NotificationsDict))
+
+    def add_message(self, keys, messages):
+        d_to_use = self
+        for key in keys:
+            d_to_use = d_to_use['children'][key]
+        if not isinstance(messages, list):
+            messages = [messages]
+        d_to_use['messages'].extend(messages)
+        return True
+
+
+@contributor_removed.connect
+def remove_contributor_from_subscriptions(contributor, node):
+    node_subscriptions = get_all_node_subscriptions(contributor, node)
+    for subscription in node_subscriptions:
+        subscription.remove_user_from_subscription(contributor)
+
+@node_deleted.connect
+def remove_subscription(node):
+    Subscription.remove(Q('object_id', 'eq', node._id))
+
+
+def get_configured_projects(user):
+    configured_project_ids = []
+    user_subscriptions = get_all_user_subscriptions(user)
+    for subscription in user_subscriptions:
+        node = Node.load(subscription.object_id)
+        if node and node.project_or_component == 'project' and not node.is_deleted and subscription.object_id not in configured_project_ids:
+            configured_project_ids.append(subscription.object_id)
+
+    return configured_project_ids
+
+
+def get_all_user_subscriptions(user):
+    user_subscriptions = []
+    for notification_type in settings.NOTIFICATION_TYPES:
+        if getattr(user, notification_type, []):
+            for subscription in getattr(user, notification_type, []):
+                if subscription:
+                    user_subscriptions.append(subscription)
+
+    return user_subscriptions
+
+
+def get_all_node_subscriptions(user, node, user_subscriptions=None):
+    if not user_subscriptions:
+        user_subscriptions = get_all_user_subscriptions(user)
+    node_subscriptions = []
+    for s in user_subscriptions:
+        if s.object_id == node._id:
+            node_subscriptions.append(s)
+
+    return node_subscriptions
+
+
+def format_data(user, node_ids, data, subscriptions_available=settings.SUBSCRIPTIONS_AVAILABLE):
```

Comment:
Can you break this into smaller functions? Particularly, the logic to serialize nodes and events can be separated. This can go a long way to make the code easier to read, maintain, and test.